### ISSUE TITLE
LSP: Language Injection/Embedding Support

### ISSUE THREAD

**Issue by JeanMertz** [neovim/neovim#26783]
### Problem

Given a document of a given file type (e.g. `html`, `markdown` or `nix`), inside this document, regions of text might not represent the ft of the source document, but are instead embedded languages within. For an HTML document, this could be inline CSS, for a Markdown document this can be any code block with a language denotation, or in nix documents (or any other generic file type with string support), an arbitrary string could be defined as needing to be parsed as a specific language.

For Tree-sitter, this already works in Neovim (I'm running `NVIM v0.10.0-dev-7f9fc2f`), by using the concept of [Language Injection](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#language-injection). Tree-sitter allows injected languages to be detected by specific rules (such as provided by [hmts.nvim](https://github.com/calops/hmts.nvim)), or by prefixing a string with a generic comment, such as `foo = /* lua */ "local bar = 'baz'"`.

While this is great, and many features and plugins (e.g. highlighting, textobjects, etc.) work with these injections out-of-the-box, it would be even better if LSP would also support a similar feature.

For example, given the [`vim-illuminate`](https://github.com/RRethy/vim-illuminate) plugin, it supports multiple "providers", `lsp`, `treesitter` and `regex`. If I enable the first two, and it detects that the root language type of the document has a configured LSP client, it will enable LSP, but I don't have the option to run a different LSP for injected languages, if I set `treesitter` as the only provider, it works as expected, and both the document and injected language types are supported by the plugin.

There are similar requests for, e.g. `conform.nvim` (see: https://github.com/stevearc/conform.nvim/issues/200), of being able to run LSPs on injected languages.

### Expected behavior

It looks like VSCode has support for this, in the form of [embedded languages](https://code.visualstudio.com/api/language-extensions/embedded-languages). Specifically, they support two different modes:

- **[Language Services](https://code.visualstudio.com/api/language-extensions/embedded-languages#language-services)**, which requires a language server to embed one or more language services to detect language injections and handle them accordingly. This seems mostly geared towards one or more languages that usually go together, think of an HTML language Server that can also handle inline CSS or JavaScript.
- **[Request Forwarding](https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding)** which seems to allow the client to determine which server should handle which part of the document, and send the request to the correct server accordingly.

I couldn't find any other issue discussing this feature, and given that NeoVim already supports Tree-sitter's language injection, this seemed like a natural (although presumably not easy) extension to the editor.

**Comment by clason** [comment]
Please lobby MSFT to include this into the open LSP standard; then we will consider implementing it. We will _not_ implemented off-standard VS Code specific features. (Tree-sitter is a different beast since there injection _is_ built into the standard.)

**Comment by lucario387** [comment]
No, saying that VSCode actually has support for embedded languages from the list they provided is just plain lying. By using the same language servers as VSCode (Volar, tsc, vscode-html-language-server), we also get all of those behaviors.

For Vue, the reason why autocompletion inside interpolations work is because the language server itself detect that and consider it to be javascript/typescript code. Another reason is because Volar is a self-replacement for typescript language server, so you are having 2 language servers in one.

Same for JavaScript (also for tree-sitter javascript), they just consider JavaScript to be a JSX file :).

HTML and PHP are also the same so I won't dive deeper here.

**Comment by JeanMertz** [comment]
> No, saying that VSCode actually has support for embedded languages from the list they gave in the link you provided above is just not right.

Fair point. As I said, it _seemed_ to me that VSCode supported the feature (I don't have VSCode installed).

> For Vue, the reason why autocompletion inside interpolations work is because the language server itself detect that and consider it to be javascript/typescript code.

That makes sense.

So yes, it seems to me that there are two use-cases:

1. Languages that are commonly used together (e.g. HTML and CSS) are handled by a single language server.
2. Arbitrary languages injected into a document with a different "root" language cannot be handled by a single language server, and instead requires a feature similar to Tree-sitter's language injection.

I'm asking for the second feature. The first seems to work already, given that this is (completely?) handled by the individual language servers.

**Comment by clason** [comment]
> I'm asking for the second feature. 

And that is a "No" until this is an official LSP feature (which it isn't).

**Comment by lucario387** [comment]
> Arbitrary languages injected into a document with a different "root" language cannot be handled by a single language server, and instead requires a feature similar to Tree-sitter's language injection.

All the languages that VSCode showed in the list are commonly used with one another. So VSCode is also not handling anything off-spec here, just sending a text range and let the language server handle all of that. Which also means VSCode cannot actually handle arbitrary injected languages, because it's fully dependent on the language server.

**Comment by JeanMertz** [comment]
> And that is a "No" until this is an official LSP feature (which it isn't). Since LSP is a "project-centric" protocol, "language injection" is simply not a thing.

Makes sense, and yes, I did get the message from your initial post. Happy to close this if you don't want any lingering feature requests that can't/won't (currently) be supported, or keep this open to gauge support and potentially have someone (likely not me, though), champion this becoming an official protocol extension.

**Comment by gwerbin** [comment]
> > Arbitrary languages injected into a document with a different "root" language cannot be handled by a single language server, and instead requires a feature similar to Tree-sitter's language injection.
> 
> All the languages that VSCode showed in the list are commonly used with one another. So VSCode is also not handling anything off-spec here, just sending a text range and let the language server handle all of that. Which also means VSCode cannot actually handle arbitrary injected languages, because it's fully dependent on a language server handling requests for multiple programming languages.

Is that also true for LSP "request forwarding" as described here? https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding

My impression is that this process is driven entirely by the client. The client detects regions of "embedded" code, then submits those regions of code to the appropriate language server. That least seems entirely doable in Neovim, as long as there's a mechanism to pass chunks of text to an arbitrary language server instance.

Bu maybe that's better suited as a 3rd-party plugin, unless/until more of a formal protocol emerges.

**Comment by clason** [comment]
> My impression is that this process is driven entirely by the client. In my interpretation of that page, the client detects regions of "embedded" code, then submits those regions of code to the appropriate language server. That least seems entirely doable in Neovim, as long as there's a mechanism to pass chunks of text to an arbitrary language server instance.

As I wrote, that's a different thing and already tracked elsewhere. The bottleneck for "mixed language" files is getting the right server to index the file.

**Comment by JeanMertz** [comment]
> already tracked elsewhere

I'm [having trouble finding this tracking issue](https://github.com/neovim/neovim/issues?q=lsp+multi+is%3Aopen), possibly because I'm not using the right terminology. Would you happen to have a link for me to read up on/start tracking?

**Comment by clason** [comment]
There's no single tracking issue; it's a planned (significant) refactor to handle multiple attached servers gracefully. But again, this will not address your original use case since in general language servers cannot attach to "parts of files".

**Comment by clason** [comment]
That's fair; I _am_ talking about 3 (which is in scope) and saying that it does not give you 2 (which is _not_ in scope, as that is something off-spec VS Code does). I am also saying that 1 and 2 are not distinct, since both require the server to be able to handle "chunks of files" (which, again, is not something the LSP specification covers, so changing that would be the -- only -- way forward).

**Comment by gwerbin** [comment]
Understood and thanks for clarifying.

It sounds like VS Code actually just uses a dirty hack for this, rather than extending the LSP protocol:

> The virtual text document provider provides CSS language server with virtual content, where all non-CSS code is replaced with whitespace.

But again, maybe that's better left for a third-party plugin.

It doesn't seem too outlandish to extract the embedded code to a new buffer in a split or floating window and attach an LSP server thereto. The UX wouldn't be too different from https://github.com/chrisbra/NrrwRgn. If anyone is looking for a project, that might be an interesting proof of concept (and a good use of Tree Sitter language injections).

**Comment by clason** [comment]
> It sounds like VS Code actually just uses a dirty hack for this, rather than extending the LSP protocol:

Exactly.

> But again, maybe that's better left for a third-party plugin.

If everybody just lets VS Code run rough shod over "its own" protocol, nothing will change. It's better to open an issue at https://github.com/microsoft/language-server-protocol and lobby for this to be standardized.

**Comment by JeanMertz** [comment]
> If everybody just lets VS Code run rough shod over "its own" protocol, nothing will change. It's better to open an issue at [microsoft/language-server-protocol](https://github.com/microsoft/language-server-protocol) and lobby for this to be standardized.

I'd be curious to know _what_ would need to be standardized in this case. Is it, support for handling regions of text instead of full document processing?

**Comment by clason** [comment]
Yes; exactly what you are asking for here.

**Comment by gegoune** [comment]
> If anyone is looking for a project, that might be an interesting proof of concept (and a good use of Tree Sitter language injections).

There already is AckslD/nvim-FeMaco.lua, which does what you are asking for but with additional step rather than somehow transparently/automatically.

**Comment by mfussenegger** [comment]
One of the main issues here is that most language servers assume they can access the file on the filesystem directly too and most also need some sort of project setup information - e.g cargo files. Because of that, to implement this without protocol extension, neovim would need to create some sort of virtual filesystem with dedicated files for the embedded regions.

And in some cases that would require language specific knowledge. For example for Java you need to have file names that match class names and directory structures that match package definitions.


Having some sort of first class document region handling in the specification would avoid all that - with the downside that servers would need to add support for it.

**Comment by justinmk** [comment]
> it's a planned (significant) refactor to handle multiple attached servers gracefully

Some discussion in https://github.com/neovim/neovim/issues/26278

**Comment by jakubbortlik** [comment]
> Understood and thanks for clarifying.
> 
> It sounds like VS Code actually just uses a dirty hack for this, rather than extending the LSP protocol:
> 
> > The virtual text document provider provides CSS language server with virtual content, where all non-CSS code is replaced with whitespace.
> 
> But again, maybe that's better left for a third-party plugin.
> 
> It doesn't seem too outlandish to extract the embedded code to a new buffer in a split or floating window and attach an LSP server thereto. The UX wouldn't be too different from https://github.com/chrisbra/NrrwRgn. If anyone is looking for a project, that might be an interesting proof of concept (and a good use of Tree Sitter language injections).

If I'm not mistaken, the [otter](https://github.com/jmbuhr/otter.nvim) plugin uses a similar trick with hidden buffers & replacing any text other than the embedded region with whitespace.

**Comment by Skyppex** [comment]
This would be a great feature when it's actually part of the LSP spec.

I work in C# .NET with GraphQL and Rider has a rather neat feature when you have the GraphQL extension installed where you just write a comment above a string like this:
```csharp
// lang=graphql
"""
// some graphql stuff
"""
```
And the GraphQL LSP actually give completion in the string based on all *.graphql files that exists in the project.
I'm certainly loving that part of Rider when i need it, but would love to have it for Neovim as it's my main editor these days.

**Comment by clason** [comment]
> This would be a great feature when it's actually part of the LSP spec.

Then you should request it to be added to the spec; commenting _here_ will not help in any way.
